### PR TITLE
Additional utility RPCs for PSBT

### PR DESCRIPTION
**Pull request by achow101**
This PR adds 3 new utility RPCs for interacting with PSBTs.

`utxoupdatepsbt` updates a PSBT with UTXO information from the node. It only works with witness UTXOs because full transactions (as would be needed for non-witness UTXOs) are not available unless txindex is enabled.

`joinpsbts` joins the inputs from multiple distinct PSBTs into one PSBT. e.g. if PSBT 1 has inputs 1 and 2, and PSBT 2 has inputs 3 and 4, `joinpsbts` would create a new PSBT with inputs 1, 2, 3, and 4.

`analyzepsbt` analyzes a PSBT and determines the current state of it and all of its inputs, and the next step that needs to be done.

### CODE REVIEW

**Comment by gmaxwell** [comment]
>  It only works with witness UTXOs because full transactions (as would be needed for non-witness UTXOs) are not available unless txindex is enabled.

It doesn't look in the wallet?

**Comment by sipa** [comment]
@gmaxwell `walletprocesspsbt` already exists for that. This is a node RPC that works without a wallet.

**Comment by achow101** [comment]
Rebased

**Comment by achow101** [comment]
Rebased

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -1791,6 +1792,320 @@ UniValue converttopsbt(const JSONRPCRequest& request)
     return EncodeBase64((unsigned char*)ssTx.data(), ssTx.size());
 }
 
+UniValue utxoupdatepsbt(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 1)
+        throw std::runtime_error(
+                            "utxoupdatepsbt \"psbt\"\n"
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Could reduce indentation?

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -1791,6 +1792,320 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+    // Clear signatures from all inputs
+    for (auto& input : merged_psbt.inputs) {
+        input.partial_sigs.clear();
+    }
+
+    // Merge
+    for (auto it = std::next(psbtxs.begin()); it != psbtxs.end(); ++it) {
+        for (unsigned int i = 0; i < it->tx->vin.size(); ++i) {
+            if (!merged_psbt.AddInput(it->tx->vin[i], it->inputs[i])) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "An input exists in multiple PSBTs");
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Could be more informative?

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -1791,6 +1792,320 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+            }
+        }
+        for (unsigned int i = 0; i < it->tx->vout.size(); ++i) {
+            merged_psbt.AddOutput(it->tx->vout[i], it->outputs[i]);
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Should allow duplicate outputs? Or should sum values into one output?

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -1791,6 +1792,320 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, strprintf("TX decode failed %s", error));
+        }
+        psbtxs.push_back(psbtx);
+    }
+
+    PartiallySignedTransaction merged_psbt(psbtxs[0]); // Copy the first one
+
+    // Clear signatures from all inputs
+    for (auto& input : merged_psbt.inputs) {
+        input.partial_sigs.clear();
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: So this is necessary because the first is a copy and the remaining psbt inputs are cleared because of `AddInput`. Maybe remove this "optimization" and merge all psbts to an empty psbt?

File: src/rpc/rawtransaction.cpp [hunk 5]
```diff
@@ -1791,6 +1792,320 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+                            "\nExamples:\n"
+                            + HelpExampleCli("joinpsbts", "\"psbt\"")
+                            );
+
+    RPCTypeCheck(request.params, {UniValue::VARR}, true);
+
+    // Unserialize the transactions
+    std::vector<PartiallySignedTransaction> psbtxs;
+    UniValue txs = request.params[0].get_array();
+    for (unsigned int i = 0; i < txs.size(); ++i) {
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Should force `txs.size() > 1`?

File: src/rpc/rawtransaction.cpp [hunk 6]
```diff
@@ -1791,6 +1792,320 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+        CAmount out_amt = 0;
+        for (const CTxOut& out : psbtx.tx->vout) {
+            out_amt += out.nValue;
+        }
+
+        // Get the fee
+        CAmount fee = in_amt - out_amt;
+
+        // Estimate the size
+        size_t size;
```
Comment (practicalswift) [inline src/rpc/rawtransaction.cpp]: The scope of `size` could be reduced?

File: src/rpc/rawtransaction.cpp (hunk 1 above)
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Done

File: src/rpc/rawtransaction.cpp (hunk 5 above)
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Done

File: src/rpc/rawtransaction.cpp (hunk 6 above)
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Done

**Comment by achow101** [comment]
Rebased

File: test/functional/rpc_psbt.py [hunk 7]
```diff
@@ -7,6 +7,7 @@
 
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import assert_equal, assert_raises_rpc_error, find_output
+from decimal import Decimal
```
Comment (practicalswift) [inline test/functional/rpc_psbt.py]: Nit: Sort imports :-)

File: src/rpc/rawtransaction.cpp [hunk 8]
```diff
@@ -1693,6 +1694,324 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+        if (input.final_script_sig.empty() && input.final_script_witness.IsNull()) {
+            input_univ.pushKV("is_final", false);
+            all_final = false;
+
+            // Figure out what is missing
+            std::vector<CKeyID> missing_pubkeys;
+            std::vector<CKeyID> missing_sigs;
+            uint160 missing_redeem_script;
+            uint256 missing_witness_script;
+            SignatureData sigdata;
```
Comment (luke-jr) [inline src/rpc/rawtransaction.cpp]: Unused?

File: src/rpc/rawtransaction.cpp [hunk 9]
```diff
@@ -1693,6 +1694,324 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+
+        // Estimate the size
+        CMutableTransaction mtx(*psbtx.tx);
+        CCoinsView view_dummy;
+        CCoinsViewCache view(&view_dummy);
+        bool success = true;
+
+        for (unsigned int i = 0; i < psbtx.tx->vin.size(); ++i) {
+            PSBTInput& input = psbtx.inputs[i];
+            SignatureData sigdata;
```
Comment (luke-jr) [inline src/rpc/rawtransaction.cpp]: Unused?

File: src/rpc/rawtransaction.cpp [hunk 10]
```diff
@@ -1693,6 +1694,324 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+    }
+    result.pushKV("inputs", inputs_result);
+
+    if (all_final) {
+        result.pushKV("next", "extractor");
+    }
+    if (calc_fee) {
+        // Get the output amount
+        CAmount out_amt = 0;
+        for (const CTxOut& out : psbtx.tx->vout) {
```
Comment (practicalswift) [inline src/rpc/rawtransaction.cpp]: Could use `std::accumulate`? :-)

**Comment by achow101** [comment]
Rebased

File: src/rpc/rawtransaction.cpp (hunk 2 above)
Comment (promag) [reply src/rpc/rawtransaction.cpp]: Like the conflicting input index?

File: src/rpc/rawtransaction.cpp (hunk 3 above)
Comment (promag) [reply src/rpc/rawtransaction.cpp]: @achow101 could you ack/nack on #12419, esp @MeshCollider https://github.com/bitcoin/bitcoin/pull/12419#issuecomment-437636604

**Review by promag — COMMENTED** [review]
Could have release note.

File: src/rpc/rawtransaction.cpp (hunk 4 above)
Comment (gwillen) [reply src/rpc/rawtransaction.cpp]: It would be less code, and less complexity, which I would say is always better, absent a reason to write more code. Why write more code?

Comment (gwillen) [reply src/rpc/rawtransaction.cpp]: (Also, this is copied from combinepsbt, where it is easiest to treat the first tx specially, because we need something to compare all the others to, to make sure the underlying transaction is the same. That restriction doesn't exist here, so the need for the extra step is gone.)

File: src/rpc/rawtransaction.cpp [hunk 11]
```diff
@@ -1773,6 +1774,324 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+            "      \"missing\" : {               (json object, optional) Things that are missing that are required to complete this input\n"
+            "        \"pubkeys\" : [             (array)\n"
+            "          \"keyid\"                 (string) Public key ID, hash160 of the public key, of a public key whose BIP 32 derivation path is missing\n"
+            "        ]\n"
+            "        \"signatures\" : [          (array)\n"
+            "          \"keyid\"                 (string) Public key ID, hash160 of the public key, of a public key whose signature is missing\n"
+            "        ]\n"
+            "        \"redeemscript\" : \"hash\"   (string) Hash160 of the redeemScript that is missing\n"
+            "        \"witnessscript\" : \"hash\"  (string) SHA256 of the witnessScript that is missing\n"
+            "        \"next\" : \"role\"           (string) Role of the next person that this input needs to go to\n"
```
Comment (gwillen) [inline src/rpc/rawtransaction.cpp]: This doesn't match the code, right? "next" goes inside the input, not inside "missing"?

**Comment by gwillen** [comment]
It looks like I'm going to depend on this for my offline-signing GUI work, so we can display nice messages in the GUI about the status of PSBTs that users hand us. Can I therefore encourage its revival? :-) @achow101

File: src/rpc/rawtransaction.cpp [hunk 12]
```diff
@@ -1773,6 +1773,67 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+            + HelpExampleCli("utxoupdatepsbt", "\"psbt\"")
+            );
+
+    RPCTypeCheck(request.params, {UniValue::VSTR}, true);
+
+    // Unserialize the transactions
+    PartiallySignedTransaction psbtx;
+    std::string error;
+    if (!DecodePSBT(psbtx, request.params[0].get_str(), error)) {
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, strprintf("TX decode failed %s", error));
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: `DecodePSBT` has its own tests.

File: src/rpc/rawtransaction.cpp [hunk 13]
```diff
@@ -1773,6 +1773,67 @@ UniValue converttopsbt(const JSONRPCRequest& request)
     return EncodeBase64((unsigned char*)ssTx.data(), ssTx.size());
 }
 
+UniValue utxoupdatepsbt(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 1)
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Put the bracket where?

**Comment by achow101** [comment]
Rebased and addressed comments.

File: src/script/sign.cpp [hunk 14]
```diff
@@ -629,6 +679,18 @@ bool PSBTInput::IsSane() const
     return true;
 }
 
+bool PSBTInput::GetUTXO(CTxOut& utxo, int prevout_index) const
```
Comment (gwillen) [inline src/script/sign.cpp]: This interface feels fragile to me. Could this instead be "bool PartiallySignedTransaction::GetInputUTXO(CTxOut& utxo, int input_index)"? That version would be resistant to misuse, whereas this version will do non-obvious crazy things if you accidentally give it the wrong prevout_index for your input (which seems easy to do.)

File: src/rpc/rawtransaction.cpp [hunk 15]
```diff
@@ -1739,6 +1741,334 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+        UniValue missing(UniValue::VOBJ);
+
+        // Check for a UTXO
+        CTxOut utxo;
+        if (input.GetUTXO(utxo, psbtx.tx->vin[i].prevout.n)) {
+            in_amt += utxo.nValue;
+            input_univ.pushKV("has_utxo", true);
+        } else {
+            input_univ.pushKV("has_utxo", false);
+            input_univ.pushKV("is_final", false);
```
Comment (gwillen) [inline src/rpc/rawtransaction.cpp]: Are there cases where this isn't overwritten by a later check? If so, you should set all_final = false here. (If not you should remove this.)

Comment (gwillen) [reply src/rpc/rawtransaction.cpp]: Oh, I see, it's handled by calc_fee = false? The flow is a little hard to follow.

**Comment by gwillen** [comment]
utACK modulo comments above. I have a general sense that analyzepsbt is a little hard to follow the logic of, but I don't immediately have suggestions for improving that.

File: src/rpc/rawtransaction.cpp [hunk 16]
```diff
@@ -1739,6 +1739,68 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+
+        if (input.non_witness_utxo || !input.witness_utxo.IsNull()) {
+            continue;
+        }
+
+        const Coin& coin = view.AccessCoin(psbtx.tx->vin[i].prevout);
+
+        std::vector<std::vector<unsigned char>> solutions_data;
+        txnouttype which_type = Solver(coin.out.scriptPubKey, solutions_data);
+        if (which_type == TX_WITNESS_V0_SCRIPTHASH || which_type == TX_WITNESS_V0_KEYHASH || which_type == TX_WITNESS_UNKNOWN) {
```
Comment (sipa) [inline src/rpc/rawtransaction.cpp]: Not P2SH? There's no guarantee that P2SH outputs are witness of course, but there is no real harm in including too much.

Comment (achow101) [reply src/rpc/rawtransaction.cpp]: I think that would be a problem right now since providing a witness utxo requires a witness signature and if a P2SH output was not witness, then it would never be signed.

File: src/script/sign.cpp [hunk 17]
```diff
@@ -237,7 +248,7 @@ bool PSBTInputSigned(PSBTInput& input)
     return !input.final_script_sig.empty() || !input.final_script_witness.IsNull();
 }
 
-bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index, int sighash)
+bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index, int sighash, std::vector<CKeyID>* missing_pubkeys, std::vector<CKeyID>* missing_sigs, uint160* missing_redeem_script, uint256* missing_witness_script, bool use_dummy)
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/script/sign.cpp]: Done

File: src/script/sign.cpp (hunk 14 above)
Comment (achow101) [reply src/script/sign.cpp]: Done

File: src/rpc/rawtransaction.cpp (hunk 15 above)
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: I'm not quite sure you are asking.

Comment (gwillen) [reply src/rpc/rawtransaction.cpp]: Yeah, I think I just have trouble following the state machine of this function, which is a little complicated, and so I was confused about whether all_final was actually getting set correctly in every case. Since nobody's relying on it yet, and I haven't found any actual issues, I think it should probably go in as-is, and I may propose some refactoring when I go to actually use it.

File: src/script/sign.h [hunk 18]
```diff
@@ -572,9 +576,12 @@ struct PartiallySignedTransaction
     bool IsNull() const;
     void Merge(const PartiallySignedTransaction& psbt);
     bool IsSane() const;
+    bool AddInput(const CTxIn& txin, PSBTInput& psbtin);
+    bool AddOutput(const CTxOut& txout, const PSBTOutput& psbtout);
     PartiallySignedTransaction() {}
     PartiallySignedTransaction(const PartiallySignedTransaction& psbt_in) : tx(psbt_in.tx), inputs(psbt_in.inputs), outputs(psbt_in.outputs), unknown(psbt_in.unknown) {}
     explicit PartiallySignedTransaction(const CMutableTransaction& tx);
+    bool GetInputUTXO(CTxOut& utxo, int input_index) const;
```
Comment (laanwj) [inline src/script/sign.h]: Please add a short doxygen documentation for the new calls.
(I know, the current ones don't, but every day is a good day to get started)

Comment (laanwj) [reply src/script/sign.h]: Maybe we can avoid the input/output order completely by returning an `Optional<CTxOut>` instead. We have this in `optional.h`, after all.

File: test/functional/rpc_psbt.py [hunk 19]
```diff
@@ -290,5 +291,72 @@ def run_test(self):
...
+        assert "witness_utxo" not in decoded['inputs'][1] and "non_witness_utxo" not in decoded['inputs'][1]
+        assert "witness_utxo" not in decoded['inputs'][2] and "non_witness_utxo" not in decoded['inputs'][2]
+        updated = self.nodes[1].utxoupdatepsbt(psbt)
+        decoded = self.nodes[1].decodepsbt(updated)
+        assert "witness_utxo" in decoded['inputs'][0] and "non_witness_utxo" not in decoded['inputs'][0]
+        assert "witness_utxo" not in decoded['inputs'][1] and "non_witness_utxo" not in decoded['inputs'][1]
+        assert "witness_utxo" not in decoded['inputs'][2] and "non_witness_utxo" not in decoded['inputs'][2]
+
+        # Two PSBTs with a common input should not be joinable
+        psbt1 = self.nodes[1].createpsbt([{"txid":txid1, "vout":vout1}], {self.nodes[0].getnewaddress():10.999})
```
Comment (laanwj) [inline test/functional/rpc_psbt.py]: `Decimal("10.999")` I guess?

File: src/script/sign.h (hunk 18 above)
Comment (achow101) [reply src/script/sign.h]: Added doxygen comment

File: test/functional/rpc_psbt.py (hunk 19 above)
Comment (achow101) [reply test/functional/rpc_psbt.py]: Done

**Comment by laanwj** [comment]
very strange!-one of the travis runs seems to be failing with
```
/bin/bash: test/test_bitcoin: No such file or directory
Makefile:12966: recipe for target 'test/addrman_tests.cpp.test' failed
make[3]: *** [test/addrman_tests.cpp.test] Error 1
make[3]: *** Waiting for unfinished jobs....
/bin/bash: test/test_bitcoin: No such file or directory
Makefile:12966: recipe for target 'test/arith_uint256_tests.cpp.test' failed
make[3]: *** [test/arith_uint256_tests.cpp.test] Error 1
PASS: qt/test/test_bitcoin-qt
```

**Comment by laanwj** [comment]
utACK 0c35f0e802274c27736b16a4542ad44bc3c7f794

File: src/rpc/rawtransaction.cpp (hunk 16 above)
Comment (sipa) [reply src/rpc/rawtransaction.cpp]: I guess there's an obvious solution which can be added later, namely letting the RPC take in descriptor(s) that apply to its inputs.

File: src/rpc/rawtransaction.cpp [hunk 20]
```diff
@@ -1691,6 +1694,334 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+            input_univ.pushKV("has_utxo", true);
+        } else {
+            input_univ.pushKV("has_utxo", false);
+            input_univ.pushKV("is_final", false);
+            input_univ.pushKV("next", "updater");
+            calc_fee = false;
+        }
+
+        // Check if it is final
+        if (input.final_script_sig.empty() && input.final_script_witness.IsNull()) {
```
Comment (sipa) [inline src/rpc/rawtransaction.cpp]: Use `PSBTInputSigned` here?

Comment (sipa) [reply src/rpc/rawtransaction.cpp]: I think you may want to add a check here for whether the input UTXO is present (if not, it's possible that you first set next="updater", and then overwrite it in this conditional).

File: src/rpc/rawtransaction.cpp [hunk 21]
```diff
@@ -1691,6 +1694,334 @@ UniValue converttopsbt(const JSONRPCRequest& request)
...
+                    for (const CKeyID& pubkey : outdata.missing_sigs) {
+                        missing_sigs_univ.push_back(HexStr(pubkey));
+                    }
+                    missing.pushKV("signatures", missing_sigs_univ);
+                }
+                input_univ.pushKV("missing", missing);
+
+                // If we are only missing signatures and nothing else, then next is signer
+                if (outdata.missing_pubkeys.empty() && outdata.missing_redeem_script.IsNull() && outdata.missing_witness_script.IsNull() && !outdata.missing_sigs.empty()) {
+                    only_missing_sigs = true;
```
Comment (sipa) [inline src/rpc/rawtransaction.cpp]: I think this (cross-input) variable can't be set to true here. If this branch is executed for the last input in a transaction, the global "next" will report "signer", even if keys/script or even UTXOs are missing for other inputs.

You probably want to have the variable start at true, and then set it to false if anything anywhere more than a signature is missing.

File: src/rpc/rawtransaction.cpp (hunk 20 above)
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Done

File: src/rpc/rawtransaction.cpp (hunk 21 above)
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Done

File: src/rpc/rawtransaction.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Done

**Comment by laanwj** [comment]
re-utACK f58b1526fd6ef76cd9dcf38c28f52616a9f9a1bf

**Comment by laanwj** [comment]
utACK 540729ef4bf1b6c6da1ec795e441d2ce56a9a58b
